File: app/src/main/java/com/akshayaap/mouseremote/network/UDPReceiver.java
package com.akshayaap.mouseremote.network;

import com.akshayaap.mouseremote.util.TaskCompleteCallback;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.SocketException;

public class UDPReceiver {

    private DatagramSocket socket = null;
    private DatagramPacket packet = null;

    private boolean onerr = false;
    private IOException ex = null;
    private TaskCompleteCallback onReceived = null;


    public UDPReceiver() throws SocketException {
        socket = new DatagramSocket();
        packet = new DatagramPacket(null, 0);
        onReceived = new TaskCompleteCallback() {
            @Override
            public void complete() {
                //TODO Nothing
            }
        };
    }

    public UDPReceiver(int port) throws SocketException {
        socket = new DatagramSocket(port);
        packet = new DatagramPacket(null, 0);
        packet.setPort(port);
        this.onReceived = new TaskCompleteCallback() {
            @Override
            public void complete() {
                //TODO Nothing
            }
        };
    }

    public void receive(byte[] buff) throws IOException {
        packet.setLength(buff.length);
        packet.setData(buff);

        new Thread() {
            @Override
            public void run() {
                try {
                    socket.receive(packet);
                    onReceived.complete();
                } catch (IOException e) {
                    onerr = true;
                    UDPReceiver.this.ex = e;
                }
            }
        }.start();
        if (onerr) {
            onerr = false;
            throw this.ex;
        }
    }

    public void receive(byte[] buff, TaskCompleteCallback onReceived) throws IOException {
        TaskCompleteCallback temp = this.onReceived;
        this.onReceived = onReceived;
        receive(buff);
        this.onReceived = temp;
    }

    public void setOnReceivedCallback(TaskCompleteCallback onReceived) {
        this.onReceived = onReceived;
    }

    @Override
    protected void finalize() throws Throwable {
        super.finalize();
        socket.close();
    }

    public void close() {
        socket.close();
    }

    public DatagramPacket getPacket() {
        return packet;
    }
}

File: app/src/main/java/com/akshayaap/touchdroid/network/UDPReceiver.kt
package com.akshayaap.touchdroid.network

import com.akshayaap.touchdroid.util.TaskCompleteCallback
import java.io.IOException
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.SocketException
import kotlin.concurrent.thread

class UDPReceiver(port: Int) {

    private val socket: DatagramSocket = DatagramSocket(port)
    var packet: DatagramPacket = DatagramPacket(ByteArray(4), 4)
        private set

    @Volatile
    private var onErr = false

    @Volatile
    private var ex: IOException? = null

    private var onReceived: TaskCompleteCallback = TaskCompleteCallback { }
    var worker: Thread? = null
    var isRunning = true

    init {
        packet.port = port
    }

    @Throws(IOException::class)
    fun receive(buff: ByteArray) {
        packet.length = buff.size
        packet.data = buff

        worker = thread {
            try {
                socket.receive(packet)
                onReceived.complete()
            } catch (e: IOException) {
                onErr = true
                this.ex = e
            }
        }
        if (onErr) {
            onErr = false
            ex?.let { throw it }
        }
    }

    @Throws(IOException::class)
    fun receive(buff: ByteArray, onReceived: TaskCompleteCallback) {
        val temp = this.onReceived
        this.onReceived = onReceived
        receive(buff)
        this.onReceived = temp
    }

    fun setOnReceivedCallback(onReceived: TaskCompleteCallback) {
        this.onReceived = onReceived
    }

    fun close() {
        worker?.interrupt()
        if (!socket.isClosed) {
            socket.close()
        }
    }
}
